src/plugins/kdb/ldap/libkdb_ldap/ldap_principal2.c

```diff
@@ -1375,64 +1492,198 @@ krb5_read_tkt_policy(krb5_context context, krb5_ldap_context *ldap_context,
-                                           &kd,
-                                           &n_kd,
-                                           mkvno);
-
-        if (st != 0) {
-            const char *msg = error_message(st);
-            st = -1; /* Something more appropriate ? */
-            k5_setmsg(context, st,
-                      _("unable to decode stored principal key data (%s)"),
-                      msg);
-            goto cleanup;
-        }
-        noofkeys += n_kd;
-        tmp = key_data;
-        /* Allocate an extra key data to avoid allocating zero bytes. */
-        key_data = realloc(key_data, (noofkeys + 1) * sizeof (krb5_key_data));
-        if (key_data == NULL) {
-            key_data = tmp;
-            st = ENOMEM;
+        err = asn1_decode_sequence_of_keys(&in, &keysets[ki].key_data,
+                                           &keysets[ki].n_key_data, mkvno);
+        if (err)
             goto cleanup;
-        }
-        for (j = 0; j < n_kd; j++)
-            key_data[noofkeys - n_kd + j] = kd[j];
-        free (kd);
+
+        if (total_keys_out)
+            total_keys += keysets[ki].n_key_data;
+        ki++;
     }
 
-    entries->n_key_data = noofkeys;
+    if (total_keys_out)
+        *total_keys_out = total_keys;
+
+    *n_keysets_out = n_keys;
+    *keysets_out = keysets;
+    keysets = NULL;
+
+cleanup:
+    free(keysets);
+    return err;
+}
+
+krb5_error_code
+krb5_decode_krbsecretkey(krb5_context context, krb5_db_entry *entries,
+                         struct berval **bvalues, krb5_kvno *mkvno)
+{
+    char                        *user=NULL;
+    krb5_key_data               *key_data=NULL, *tmp;
+    krb5_error_code             err=0;
+
+    struct keyset *keys = NULL;
+    unsigned int i, n_key_entries, total_keys;
+
+    err = krb5_unparse_name(context, entries->princ, &user);
+    if (err)
+        goto cleanup;
+
+    err = decode_keys(bvalues, mkvno, &keys, &n_key_entries, &total_keys);
+    if (err != 0) {
+        const char *msg = error_message(err);
+        k5_setmsg(context, err,
+                  _("unable to decode stored principal key data (%s)"),
+                  msg);
+        goto cleanup;
+    }
+
+    key_data = k5calloc(total_keys, sizeof(krb5_key_data), &err);
+    if (key_data == NULL)
+        goto cleanup;
+
+    tmp = key_data;
+    for (i = 0; i < n_key_entries; i++) {
+        memcpy(tmp, keys[i].key_data,
+               sizeof(krb5_key_data) * keys[i].n_key_data);
+        tmp += keys[i].n_key_data;
+    }
+    entries->n_key_data = total_keys;
     entries->key_data = key_data;
 
 cleanup:
     free (user);
-    return st;
+    free (keys);
+    return err;
+}
+
+/*
+ * Sort the old keys for a principal entry by kvno in descending order. Expects
+ * each key history entry to only contain keys with the same kvno.
+ */
+static void
+sort_old_keys(osa_princ_ent_rec *adb)
+{
+    size_t counter, indx;
+    osa_pw_hist_ent tmp;
+
+    /* Use insertion sort. */
```

Comment:
Well, this works, so, unless there's some pressing need to switch to qsort I don't see the point.